nixos/modules/services/networking/wireguard.nix

```diff
@@ -0,0 +1,223 @@
+        inclusive, how often to send an authenticated empty packet to the peer, 
+        for the purpose of keeping a stateful firewall or NAT mapping valid 
+        persistently. For example, if the interface very rarely sends traffic, 
+        but it might at anytime receive traffic from a peer, and it is behind 
+        NAT, the interface might benefit from having a persistent keepalive 
+        interval of 25 seconds; however, most users will not need this.'';
+      };
+
+    };
+
+  };
+
+  generateConf = name: values: pkgs.writeText "wireguard-${name}.conf" ''
+    [Interface]
+    PrivateKey = ${values.privateKey}
+    ${optionalString (values.presharedKey != null) "PresharedKey = ${values.presharedKey}"}
+    ${optionalString (values.listenPort != null)   "ListenPort = ${toString values.listenPort}"}
+  
+    ${concatStringsSep "\n\n" (map (peer: ''
+    [Peer]
+    PublicKey = ${peer.publicKey}
+    ${optionalString (peer.allowedIPs != []) "AllowedIPs = ${concatStringsSep ", " peer.allowedIPs}"}
+    ${optionalString (peer.endpoint != null) "Endpoint = ${peer.endpoint}"}
+    ${optionalString (peer.persistentKeepalive != null) "PersistentKeepalive = ${toString peer.persistentKeepalive}"}
+    '') values.peers)}
+  '';
+
+  # Set one systemd unit per interface
+  generateUnit = name: values: 
+    nameValuePair "wireguard-${name}" 
+      {
+        description = "Wireguard unit for interface ${name}";
+        wantedBy = [ "ip-up.target" ];
+        path = with pkgs; [ iproute wireguard ];
+        serviceConfig.Type = "oneshot";
+        script = 
+          let 
+            preSetup = if values.preSetup != null
+              then values.preSetup
+              else ''
+                ip link del dev ${name} 2>/dev/null || true
+                echo "adding interface..."
+                ip link add dev ${name} type wireguard
+              '';
+            postSetup = if values.postSetup != null
+              then values.postSetup
+              else ''
+                echo "adding ip addresses..."
+                ${concatStrings (map (ip: ''
+                ip address add ${ip} dev "${name}"
+                '') values.ips)}
+
+                echo "setting the interface up..."
+                ip link set up dev "${name}"
+
+                echo "adding routes..."
+                ${concatStrings (flatten (map (peer: (map (ip: ''
+                ip route add ${ip} dev "${name}"
+                '') peer.allowedIPs)) values.peers))}
+              '';
+            in ''
+              ${preSetup}
+              echo "configuring interface..."
+              wg setconf "${name}" ${generateConf name values}
+              ${postSetup}
+            '';
+      };
+
+in
+
+{
+
+  ###### interface
+
+  options = {
+
+    networking.wireguard = {
+
+      interfaces = mkOption {
+        description = "Wireguard interfaces.";
+        default = {};
+        # TODO add useful example when the module interface is stabilized
```

Comment:
Seems better to add a useful interface before the module interface is stabilized.